### DISCUSSION TITLE
Is it possible to serve embeddings with FT backend?

### DISCUSSION THREAD

**Discussion by denisyarats** [moyix/fauxpilot#97] [Q&A]
Is it possible to configure FT backend to return hidden state for the last token in the sequence (e.g. LLM embedding) for GPT-J?
Or this backend needs to be written in TensorRT?

**Comment by moyix** [comment]
Hmm, I don't think there's any configuration option in FasterTransformer that would let you do this, it would involve modifying the code to FasterTransformer.

**Reply by denis-pplx** [reply]
what would be a good alternative then, use TensorRT?

**Reply by thakkarparth007** [reply]
We have a Python backend now, it can be modified to do this. But what's the use case?

**Reply by moyix** [reply]
Hmm, one possibility (if embeddings are fast) would be to build a database of embeddings for your current project in the IDE plugin, then at inference time, embed the current function / context and ask for the nearest neighbors in the source to decide what code should go into the prompt.

Of course this all pretty speculative; I don't know of any IDE plugins right now that are nearly this sophisticated...

**Comment by thakkarparth007** [comment]
Aah, yeah that'd be cool but also expensive (time-wise).

**Comment by thakkarparth007** [comment]
Related -- the CoCoMIC paper also talks about using embeddings but not for identifying the most relevant context, but instead for predicting the completion once the context has been collected.

<img width="615" alt="image" src="https://user-images.githubusercontent.com/747423/209556972-6ba79f68-a428-4674-92d0-50ec3442047a.png">

Basically, for each context item, they compute one embedding, independent of other items. They then use this embedding as an input to the prediction pipeline. Note that this paper performs its own training on top of codegen.
